libraries/browser-tracker-core/src/tracker/id_cookie.ts

```diff
@@ -0,0 +1,361 @@
+   */
+  userId: string;
+
+  /**
+   * An identifier for the session (same as domain_sessionid)
+   */
+  sessionId: string;
+
+  /**
+   * The index of the current session for this user (same as domain_sessionidx)
+   */
+  sessionIndex: number;
+
+  /**
+   * Index of the current event in the session
+   */
+  eventIndex: number;
+
+  /**
+   * The previous session identifier for this user
+   */
+  previousSessionId: string | null;
+
+  /**
+   * The mechanism that the session information has been stored on the device
+   */
+  storageMechanism: string;
+
+  /**
+   * Identifier of the first event for this session
+   */
+  firstEventId: string | null;
+
+  /**
+   * Date-time timestamp of when the first event in the session was tracked
+   */
+  firstEventTimestamp: string | null;
+}
+
+export function emptyIdCookie() {
+  const idCookie: ParsedIdCookie = ['1', '', 0, 0, 0, undefined, '', '', '', undefined, 0];
+  return idCookie;
+}
+
+/**
+ * Parses the cookie values from its string representation.
+ *
+ * @param id Cookie value as string
+ * @param domainUserId Domain user ID to be used in case of empty cookie string
+ * @returns Parsed ID cookie tuple
+ */
+export function parseIdCookie(
+  id: string | undefined,
+  domainUserId: string,
+  memorizedSessionId: string,
+  memorizedVisitCount: number
+) {
+  let now = new Date(),
+    nowTs = Math.round(now.getTime() / 1000),
+    tmpContainer;
+
+  if (id) {
+    tmpContainer = id.split('.');
+    // cookies enabled
+    tmpContainer.unshift('0');
+  } else {
+    tmpContainer = [
+      // cookies disabled
+      '1',
+      // Domain user ID
+      domainUserId,
+      // Creation timestamp - seconds since Unix epoch
+      nowTs,
+      // visitCount - 0 = no previous visit
+      memorizedVisitCount,
+      // Current visit timestamp
+      nowTs,
+      // Last visit timestamp - blank meaning no previous visit
+      '',
+      // Session ID
+      memorizedSessionId,
+    ];
+  }
+
+  if (!tmpContainer[sessionIdIndex] || tmpContainer[sessionIdIndex] === 'undefined') {
+    // session id
+    tmpContainer[sessionIdIndex] = uuid();
+  }
+  if (!tmpContainer[previousSessionIdIndex] || tmpContainer[previousSessionIdIndex] === 'undefined') {
+    // previous session id
+    tmpContainer[previousSessionIdIndex] = '';
+  }
+  if (!tmpContainer[firstEventIdIndex] || tmpContainer[firstEventIdIndex] === 'undefined') {
+    // firstEventId - blank meaning no previous event
+    tmpContainer[firstEventIdIndex] = '';
+  }
+  if (!tmpContainer[firstEventTsInMsIndex] || tmpContainer[firstEventTsInMsIndex] === 'undefined') {
+    // firstEventTs - blank meaning no previous event
+    tmpContainer[firstEventTsInMsIndex] = '';
+  }
+  if (!tmpContainer[eventIndexIndex] || tmpContainer[eventIndexIndex] === 'undefined') {
+    // eventIndex – 0 = no previous event
+    tmpContainer[eventIndexIndex] = 0;
+  }
+
+  const parseIntOr = (value: any, defaultValue: any) => {
```

Comment:
nit: I'd usually use `unknown` here rather than `any`.

Reply:
It actually breaks the type checking in this case, because if I change `defaultValue` to unknown, TypeScript won't automatically infer its type based on what is passed and the `parseIntOr` will always have the return type `unknown`. Instead if I leave it as any, the implicit return type of the function is inferred from the arguments I pass to it. At least that is my understanding based on the type check errors that I am getting in VS Code.

Reading [this Stackoverflow](https://stackoverflow.com/a/51439876/3334072) maybe a better explanation is that you can't assign unknown to other types (e.g., number). So I would always have to type cast the result of `parseIntOr` on the lines below where I use it.